sqlx-sqlite/src/index.rs

```diff
@@ -0,0 +1,651 @@
+                                }
+                            }
+                        }
+                        // If we have A AND B but we can't push down B we can still push down A
+                        // because A must be true for the whole expression to be true
+                        (Some(left_pushdown), None, Operator::And) => Some(left_pushdown),
+                        // Same for the other side
+                        (None, Some(right_pushdown), Operator::And) => Some(right_pushdown),
+                        _ => None
+                    }
+                }
+            }
+        },
+        Expr::Not(inner) => {
+            let inner_pushdown = push_down_filter(inner);
+            inner_pushdown.map(|inner_pushdown| inner_pushdown.not())
+        },
+        Expr::Like(inner) => {
+            let negated = inner.negated;
+            let expr = *inner.expr.clone();
+            let pattern = *inner.pattern.clone();
+            let escape_char = inner.escape_char;
+            let case_insensitive = inner.case_insensitive;
+
+            let column = match &expr {
+                Expr::Column(column) => column.clone(),
+                _ => return None
+            };
+            let pattern = match &pattern {
+                Expr::Literal(ScalarValue::Utf8(Some(pattern))) => pattern.clone(),
+                _ => return None
+            };
+            // We don't support escape characters in this example
+            if escape_char.is_some() {
+                return None;
+            }
+            // Find the prefix in pattern by looking for the first `%` and truncate
+            let prefix_len = pattern.chars().position(|c| c == '%').unwrap_or_else(|| pattern.len());
+            let mut prefix = pattern.chars().take(prefix_len).collect::<String>();
+            let mut min_val_col = SeaQExpr::col(ColumnStatistics::StringMinValue);
+            // If this is a case insensitive match we need to convert the prefix to lowercase
+            if case_insensitive {
+                prefix = prefix.to_lowercase();
+                min_val_col = SeaQExpr::expr(sea_query::Func::lower(min_val_col));
+            };
+
+            let filter = SeaQExpr::col(ColumnStatistics::ColumnName).eq(column.name).and(
+                min_val_col.gte(SqlValue::String(Some(Box::new(prefix.clone()))))
+            );
+            if negated {
+                Some(filter.not())
+            } else {
+                Some(filter)
+            }
+        },
+        // We could handle more cases here, at least simple ones involving nulls, negations, etc.
+        // But this example does not implement that
+        _ => None
+    }
+}
+
+/// Push down a simple binary expression to the index
+/// Only a subset of expressions are supported since `a = 1` has to be rewritten as `a_int_max_value >= 1 AND a_int_min_value <= 1`
+fn push_down_binary_filter(value: &ScalarValue, op: &Operator) -> Option<SimpleExpr> {
```

Comment:
It might be clearer to call this "rewrite_filter" or "rewrite_binary_expr" as it isn't really "pushing down" the filter in my mind -- it is creating  `SimpelExpr` that wll be "pushed down" (into sqlite)

Reply:
BTW this rewrite looks like a combination
1. The expr rewrite that  [PruningPredicate](https://docs.rs/datafusion/latest/datafusion/physical_optimizer/pruning/struct.PruningPredicate.html) does internally 
2. A translation from the DataFusion `Expr`s to the sea_query

If you separated the two passes the logic might be clearer and maybe you could reuse some existing code:

Some thoughts:
1. Consider using the rewritten PruningPredicate expr directly (and there has been a lot of effort in testing that rewrite for correctness) 
2. Use the `Expr` --> `SQL` code in [`expr_to_sql`](https://docs.rs/datafusion/latest/datafusion/sql/unparser/fn.expr_to_sql.html)  instead of sea_query (as the code already exists

Reply:
This is great, I was wondering if there's existing tested code to do these conversions. I see that I can build a `PruningPredicate` and call  [`PruningPredicate::predicate_expr`](https://docs.rs/datafusion/latest/datafusion/physical_optimizer/pruning/struct.PruningPredicate.html#method.predicate_expr) to get the rewritten expr so I can then run that against SQLite and not have to materialize the statistics in memory (which is what `PruningPredicate` seems to be used for generally).

However the way that rewriting happens it expects the schema of the statistics to match the schema of the table: `value` needs columns `value_min` and `value_max` of the correct type.
That makes sense when your statistics are stored with the actual data or in memory but breaks down if you use a traditional RDMS and want to support more than 1 table schema because you don't want to make a super wide table and add dynamically generated columns.

The schema I'm using here gets around that by storing a sparse table of values with `(file_id, row_group, column_name)` as the primary key. But I don't see how to massage what `PruningPredicate::predicate_expr` gives you into that without doing another rewrite pass, at which point there's less value in using `PruningPredicate::predicate_expr` right?

What do you think generally of the statistics schema being used here, is it a bad idea and we should be using something more similar to what `PruningPredicate::predicate_expr` already knows or is my intuition that you'd want something more flexible like I have here for statistics that aren't stored with the data onto something?

Reply:
Looking a bit further into what `PruningPredicate` is doing it seems to me like the `LiteralGuarantees` part is not necessary if I'm pushing work to another database. Let me know if that sounds right or not.

Reply:
I did a first attempt at making `stat_column_expr` pluggable. This was a baby step that may make sense regardless: https://github.com/apache/datafusion/pull/10863. I am a bit unsure how to proceed after that, the current API for `RequiredColumns` seems to be pretty entrenched with just replacing the column name whereas this would need to replace `a > 1` with `column_name = 'a' AND int_min > 1` instead of `a_min > 1`.

Reply:
After a think I realized that it's trivially provable that what I am doing here does not work for more than 1 column: `x = 5 AND y = 7` would become `(col = 'x' AND int_min <= 5) AND (col = 'x' AND int_max >= 5) AND (col = 'y' AND int_min <= 7) AND (col = 'y' AND int_max >= 7)` which will never match any rows because `col` can never be `5` and `7` at the same time.

Reply:
I think here you mean `col` can never be `x` and `y` at the same time 👀

Reply:
I see -- yes, if you want to push the evaluation down into the lower level database without a column per statistic (which could indeed result in a large number of columns) it would be hard to reuse the exisitng pruning predicate logic

Reply:
I think we may still be able to make it work by doing a join per column. Which I don't love, but should work and avoids any dynamic schemas in the index database. I'll give that a try and loop back.

Reply:
I'm going to rework this to be simple and hardcoded. It was getting out of hand to try to make this fancier for an example.

@alamb is there a way to go from a `PhysicalExpr` to an `Expr` if I want to use `expr_to_sql`? It's another thing where it would be simpler for the example but not universal (I assume it generates datafusion SQL, there's no guarantee that's valid for the lower level index database).

Reply:
@alamb I've reworked this to use `PruningPredicate` to do the rewriting. The statistics table now has largely the same schema as what `PruningPredicate` expects.

I did _not_ use `expr_to_sql` because that generates DataFusion flavored SQL which would not be compatible with many index databases. Instead I opted to implement a function to convert from `PhysicalExpr` -> `sea_query::SimpleExpr`. I think this is valuable to have as an example for folks, it covers a wide range of index databases (since SeaQuery can generate SQL for SQLite, Postgres, MySQL, or be extended) and is not that much code (at least for the cases I chose to handle).